### PR TITLE
Project Pipeline

### PR DESCRIPTION
**Pull request by MauricioUyaguari**
Adds project pipeline. 
See https://github.com/finos/legend-sdlc/issues/233

### CODE REVIEW

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/finos/FinosFileLoader.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.eclipse.collections.api.map.primitive.MutableIntObjectMap;
+import org.eclipse.collections.impl.factory.primitive.IntObjectMaps;
+import org.finos.legend.sdlc.server.tools.IOTools;
+import org.finos.legend.sdlc.server.tools.StringTools;
+
+import java.io.InputStream;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+
+public class FinosFileLoader
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/finos/FinosFileLoader.java:26]: This should probably be package-private.

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/finos/FinosFileLoader.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+{
+    private final MutableIntObjectMap<String> cache = IntObjectMaps.mutable.empty();
+    private final ClassLoader classLoader;
+    private final String schemaSignature;
+    private final String fileName;
+
+    FinosFileLoader(ClassLoader classLoader, String fileName, String schemaSignature)
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/finos/FinosFileLoader.java:33]: Rename "schemaSignature" to "fileExtension" to make it more clear.

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructureExtensions.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.finos.legend.sdlc.server.project;
+
+import java.util.Map;
+
+public class ProjectStructureExtensions
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructureExtensions.java:19]: Does this have anything to do with the class ProjectStructureExtension? If not, then you should find a different name for this class.

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/config/ProjectCreationConfiguration.java [hunk 4]
```diff
@@ -148,6 +148,13 @@ public Pattern getArtifactIdPattern()
         return this.artifactIdPattern;
     }
 
+
+    @JsonCreator
+    public static ProjectCreationConfiguration newConfig(@JsonProperty("defaultProjectStructureVersion") Integer defaultProjectStructureVersion, @JsonProperty("groupIdPattern") Pattern groupIdPattern, @JsonProperty("artifactIdPattern") Pattern artifactIdPattern, @JsonProperty("disallowedTypes") Collection<DisallowedType> disallowedTypes)
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/config/ProjectCreationConfiguration.java:153]: Why is this being moved away from the other factory methods?

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/config/ProjectStructureExtensionsConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.eclipse.collections.api.factory.Maps;
+import org.finos.legend.sdlc.server.project.ProjectStructureExtensions;
+
+import java.util.Collections;
+import java.util.Map;
+
+public class ProjectStructureExtensionsConfiguration
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/config/ProjectStructureExtensionsConfiguration.java:25]: See my comment on ProjectStructureExtensions.

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]
This is not a complete review, but there are several comments here.

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectConfigurationUpdateBuilder.java [hunk 6]
```diff
@@ -314,6 +324,12 @@ public ProjectConfigurationUpdateBuilder withProjectStructureVersion(Integer pro
         return this;
     }
 
+    public ProjectConfigurationUpdateBuilder withProjectExtensionsOverride(ProjectStructurePlatformExtensions projectStructurePlatformExtensions)
+    {
+        setProjectExtensionsCollections(projectStructurePlatformExtensions);
+        return this;
+    }
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectConfigurationUpdateBuilder.java:331]: Please change this method name to match whatever you change the other two to.

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructure.java [hunk 7]
```diff
@@ -320,7 +320,12 @@ public static ProjectStructure getProjectStructure(FileAccessContext fileAccessC
 
     public static ProjectStructure getProjectStructure(ProjectConfiguration projectConfiguration)
     {
-        return PROJECT_STRUCTURE_FACTORY.newProjectStructure(projectConfiguration);
+        return PROJECT_STRUCTURE_FACTORY.newProjectStructure(projectConfiguration, null);
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructure.java:323]:
```suggestion
        return getProjectStructure(projectConfiguration, null);
```

**Comment by MauricioUyaguari** [comment]
add tests for serializer dependency

File: legend-sdlc-server/src/main/resources/org/finos/legend/sdlc/server/gitlab/finos/ci-settings-1.xml [hunk 8]
```diff
@@ -0,0 +1,46 @@
+<!--
+ Copyright 2020 Goldman Sachs
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/resources/org/finos/legend/sdlc/server/gitlab/finos/ci-settings-1.xml:2]: 2021

File: legend-sdlc-server/src/main/resources/org/finos/legend/sdlc/server/gitlab/finos/gitlab-ci-1.yml [hunk 9]
```diff
@@ -0,0 +1,61 @@
+# Copyright 2020 Goldman Sachs
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/resources/org/finos/legend/sdlc/server/gitlab/finos/gitlab-ci-1.yml:1]: 2021

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]
Another incomplete review, but some comments.

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/guice/AbstractBaseModule.java [hunk 10]
```diff
@@ -218,6 +220,15 @@ private ProjectStructureConfiguration getProjectStructureConfiguration()
...
     }
 
+    private ProjectStructurePlatformExtensions getProjectStructurePlatformExtensions()
+    {
+        if (this.getProjectStructureConfiguration().getProjectPlatformsConfiguration() == null)
+        {
+            return null;
+        }
+        return this.getProjectStructureConfiguration().getProjectPlatformsConfiguration().buildProjectStructurePlatformExtensions();
+    }
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/guice/AbstractBaseModule.java:230]: Let's only compute this once and store it in a field, similar to getProjectStructureExtensionProvider.

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructurePlatformExtensions.java [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+            }
+            extensionsCollectionsMap.put(collection.name, collection);
+        });
+        return new ProjectStructurePlatformExtensions(platformsMap, extensionsCollectionsMap);
+    }
+
+    public List<Platform> getPlatforms()
+    {
+        return this.platforms.values().stream().collect(Collectors.toList());
+    }
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructurePlatformExtensions.java:66]:
```suggestion
    public List<Platform> getPlatforms()
    {
        return Lists.mutable.withAll(this.platforms.values());
    }
```

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructurePlatformExtensions.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+        public String getGroupId()
+        {
+            return groupId;
+        }
+
+        public Map<Integer, String> getProjectStructureVersions()
+        {
+            return projectStructureVersions;
+        }
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructurePlatformExtensions.java:117]: Do we need to give access to this map? If not, then remove this method. If we do, then let's make sure it's unmodifiable (e.g., by wrapping it with Collections.unmodifiableMap).

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]
Just a few minor changes

**Review by kevin-m-knight-gs — APPROVED** [review]

**Comment by akphi** [comment]
🎉 @MauricioUyaguari 🌮

**Pull request merged**